pyro/distributions/hmm.py

```diff
@@ -423,6 +424,32 @@ def filter(self, value):
         # Convert to a distribution.
         return Categorical(logits=logp, validate_args=self._validate_args)
 
+    @torch.no_grad()
+    def sample(self, sample_shape=torch.Size()):
+        assert self.duration is not None
+
+        # Sample initial state.
+        S = self.initial_logits.size(-1)  # state space size
+        init_shape = torch.Size(sample_shape) + self.batch_shape + (S,)
+        init_logits = self.initial_logits.expand(init_shape)
+        x = Categorical(logits=init_logits).sample()
+
+        # Sample hidden states over time.
+        trans_shape = self.batch_shape + (self.duration, S, S)
+        trans_logits = self.transition_logits.expand(trans_shape)
+        xs = []
+        for t in range(self.duration):
+            x = Categorical(logits=Vindex(trans_logits)[..., t, x, :]).sample()
+            xs.append(x)
+        x = torch.stack(xs, dim=-1)
+
+        # Sample observations conditioned on hidden states.
+        obs_shape = self.batch_shape + (self.duration, S)
```

Comment:
shouldn't we index and then sample?

Reply:
I've added a comment explaining why the efficient approach is hard (see also #3052).